pkg/frontend/querymiddleware/split_by_instant_interval.go

```diff
@@ -0,0 +1,106 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package querymiddleware
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/grafana/dskit/tenant"
+	"github.com/prometheus/prometheus/promql"
+	"github.com/prometheus/prometheus/promql/parser"
+
+	apierror "github.com/grafana/mimir/pkg/api/error"
+	"github.com/grafana/mimir/pkg/frontend/querymiddleware/astmapper"
+	"github.com/grafana/mimir/pkg/storage/lazyquery"
+)
+
+// TODO: add metrics
+
+// splitByInstantIntervalMiddleware is a Middleware that can (optionally) split the instant query by splitInterval
+type splitByInstantIntervalMiddleware struct {
+	next   Handler
+	limits Limits
+	logger log.Logger
+
+	engine *promql.Engine
+
+	splitEnabled  bool
+	splitInterval time.Duration
+}
+
+// newSplitByInstantIntervalMiddleware makes a new splitByInstantIntervalMiddleware.
+func newSplitByInstantIntervalMiddleware(
+	splitEnabled bool,
+	splitInterval time.Duration,
+	limits Limits,
+	logger log.Logger,
+	engine *promql.Engine) Middleware {
+
+	return MiddlewareFunc(func(next Handler) Handler {
+		return &splitByInstantIntervalMiddleware{
+			splitEnabled:  splitEnabled,
+			next:          next,
+			limits:        limits,
+			splitInterval: splitInterval,
+			logger:        logger,
+			engine:        engine,
+		}
+	})
+}
+
+func (s *splitByInstantIntervalMiddleware) Do(ctx context.Context, req Request) (Response, error) {
+	_, err := tenant.TenantIDs(ctx)
```

Comment:
Create a tracing span as first thing, for example:
```
log, ctx := spanlogger.NewWithLogger(ctx, s.logger, "splitByInstantIntervalMiddleware.Do")
defer log.Span.Finish()
```

Reply:
Done 👍